cEP-0022.md

```diff
@@ -0,0 +1,162 @@
+<https://github.com/processing/processing/pull/4781>.
+
+# Implementation
+
+This project will be implemented in two phases:
+
+- The Brute Force way
+- Further optimization on the Brute Force
+
+1. The first step will consist of preparing a dict of all the values that can
+   be attained by the bears' settings, may it be necessary or optional as we need
+   to create *green* config files that are as specific as possible. Failure to do
+   so will lead to dropping of most **optional** bear settings, thus will lead to
+   coala not detecting many inconsistencies, and the developers of the project will
+   have to do the work of learning about the bears and their settings and then
+   modifying the `.coafile` manually, while not able to guess the correct
+   **necessary** bear settings will lead to dropping of bears themselves.
+
+**The following steps mainly describe the brute force way**
+
+2. Certain thresholds may be defined for an org based on several factors
+   for the number of inconsistencies detected to be considered mistakes however
+   these shall be stored for later as the first and foremost goal of this project
+   is to generate completely *"green"* config files.
+
+3. coala will be ran over and over again for the entire project with all
+   instances of bears with all the combinations of settings running in
+   parallel, as soon as instances of a particular setting is done and we get
+   a value for the setting for which no errors were generated, it goes directly
+   into the `.coafile`, if no value for an optional setting matches, the setting will
+   have to be dropped,  while if no necessary setting is found out to be *green*,
+   the entire bear will have to be dropped, keeping at the back (of our heads) that,
+   certain settings producing number of errors within the threshold as described in
+   step 2 are being saved for later.
+
+4. Once we have collected this data and using the metadata already being
+   collected from `coala-quickstart` we will generate our first and foremost complete
+   *green* config file.
+
+5. The additional data generated for erraneous settings within a theshold can
+   be reported later to the project administrators in any form and wait for their
+   decision whether these are actual codestyles or mistakes. In case of mistakes,
+   `coala-quickstart` should have the ability to add these rules automatically
+   to the config file. In case these are actual codestyles, `coala-quickstart`
+   should store this data for use in the next phase.
+
+# Improving upon the Brute Force
+
+Premature optimization should not be performed on anything
+although some of the limitations of the brute force method are clear even
+at the beginning. We try to rush through the brute force as quick as possible
+and in the mean time keep on thinking of enhancing this in ways to detect more
+specific bear settings and reducing the total run time of itself.
+
+## The *"lite-mode"*
```

Comment:
So, instead of running on "all the files of particular type", we are running on "few randomly drawn files of a particular type", right?

Sounds like a nice approach.

Reply:
yess

Reply:
I dont see global bears mentioned in this cEP.

They need to be excluded at least from this lite mode obviously, and probably treated differently in most of the algorithms, at least delayed until after a stable combination of local bear settings has been found.